### ISSUE TITLE
Add example using grpc and http2 

### ISSUE THREAD

**Issue by bowei** [kubernetes/ingress-gce#18]
_From @aledbf on December 1, 2016 22:39_

_Copied from original issue: kubernetes/ingress-nginx#39_

**Comment by bowei** [comment]
_From @aledbf on December 30, 2016 3:24_

Something like this should work:

https://github.com/caiofilipini/grpc-weather/

- create deployment
```
kubectl run \
  --image=caiofilipini/grpc-weather:master grpc-weather \
  --port=9000 \
  --env="OPEN_WEATHER_MAP_API_KEY=<token from http://openweathermap.org/api>" \
  --env="WEATHER_UNDERGROUND_API_KEY=<token from http://openweathermap.org/api>"
```

- expose service
```
kubectl expose deployment grpc-weather --port=9000 --name=grpc-weather
```

- create ingress
```
echo "
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: grpc-weather
spec:
  rules:
  - host: grpc-weather
    http:
      paths:
      - backend:
          serviceName: grpc-weather
          servicePort: 9000
        path: /
" | kubectl create -f -
```

- use docker image to run the client
```
docker run \
  --rm \
  -it \
  --name weather_service \
  --entrypoint bash \
  --net=host \
  caiofilipini/grpc-weather:master
```

- execute
```
echo "<grpc-weather svc IP> grpc-weather" >>/etc/hosts
curl -v <grpc-weather svc IP>:80 -H 'Host: grpc-weather"
make build-client
./weather_client/client --s grpc-weather --p 80 Santiago
```

**Comment by bowei** [comment]
_From @krancour on April 3, 2017 16:21_

@aledbf I haven't tried, but does this work? I recently tried using a different upstream grpc service and requests kept failing because of connection reset by peer.

**Comment by bowei** [comment]
_From @aledbf on April 3, 2017 16:25_

@krancour works if you use the `ssl-passthrough` annotation, i.e. just use nginx for the TLS hello and then terminate SSL in the pod (this is because nginx does not support http2 in the upstream)

**Comment by bowei** [comment]
_From @krancour on April 3, 2017 16:27_

@aledbf ahhh... thanks! I will try that.

**Comment by bowei** [comment]
_From @philipithomas on June 28, 2017 21:58_

can a grpc server listen on port 80? More specifically - how can ssl-passthrough be configured for port 80?

**Comment by bowei** [comment]
_From @aledbf on June 28, 2017 21:59_

@philipithomas I just answered this in your issue :)

**Comment by bowei** [comment]
_From @drigz on August 17, 2017 9:8_

For others following this trail, the other issue is #923.

**Comment by shankj3** [comment]
Any luck getting this to work, @bowei ? I am having the same issue.

**Comment by munjalpatel** [comment]
I am also interested in exposing internal `gRPC` based services to external clients through GKE ingress controller.

**Comment by bgetsug** [comment]
Is there anyone who can speak to the priority of completing this or anyone who can currently provide a working example since #146 has been merged?

I am struggling to configure HTTPS (HTTP/2) ingress on GKE in front of the Cloud Endpoints ESP for a gRPC service. To clarify I'm trying to achieve the following architecture:

```
GCP HTTPS LB (includes HTTP/2 ?)
         |
Cloud Endpoints ESP listenting for HTTP/2, deployed to GKE, exposed as NodePort Service
         |
gRPC server deployed to GKE
``` 

Anything seem wrong or unreasonable with that?

I have referred to the following existing documentation/samples:

- https://github.com/GoogleCloudPlatform/golang-samples/tree/master/endpoints/getting-started-grpc
- https://github.com/GoogleCloudPlatform/endpoints-samples/tree/master/k8s
- https://github.com/ahmetb/gke-letsencrypt

At most, I am able to get ESP working with my gRPC service and can connect via the pod IPs on the private network. The health check created by the ingress does not seem to be working. Is there any way to configure it for HTTP/2 or TCP instead of HTTP?

I would also welcome any alternative suggestions including manually creating an LB or anything I might orchestrate with Terraform. Thank you.

**Comment by bgetsug** [comment]
Well after much knob twiddling I was able to access a gRPC+Cloud Endpoints (ESP) service through an HTTPS LB. I ended up configuring everything manually via the web console, but will probably write a script so I can at least replicate the process with `gcloud`.

A few notes that may help others:
- https://cloud.google.com/load-balancing/docs/https/#components
  > To use gRPC with your Google Cloud Platform applications, you must proxy requests end-to-end over HTTP/2. To do this with an HTTP(S) load balancer: 
  > `1. Configure an HTTPS load balancer.`
  > `2. Enable HTTP/2 as the protocol from the load balancer to the backends.`
  > The load balancer negotiates HTTP/2 with clients as part of the SSL handshake by using the ALPN TLS extension.
- https://cloud.google.com/load-balancing/docs/backend-service#HTTP2-limitations
  > HTTP/2 to backends is not supported for Kubernetes Engine

    Does that mean it doesn't work, or just not "supported"?

- Configuring the LB backend to use ESP's HTTP/2 port did not work. I had to enable the SSL port, mount the certs (same as the LB frontend) to the container (`Deployment`), and expose it via the `Service`. So the backend is configured for the HTTP/2 protocol, but points to ESP's SSL port. 
- I enabled ESP's `/healthz` endpoint and used that for the backend health check, with the same protocol and port as the backend.

Unfortunately, I was never able to make an `Ingress` behave due to the inability to customize backend health checks. For that, it sounds like we'll need resolution on #42.

**Comment by manjotpahwa1** [comment]
Not yet, in progress.

**Comment by clehene** [comment]
@bgetsug I was able to override the LB health check to point to the ESP http_port and the probe will see it healthy. However it gets overridden (I guess by LGBC) after a while back to the main service port (HTTP/2).

I assume GLBC sets the probe to Ingress/spec/backend/servicePort and, hence, any other probe will not really work. 

I was not able to figure out how the health probe is set by GLBC. After shaving yaks trying to get GLBC which uses non standard vendoring paths and GoLand's stubbornness in not accepting anything else but `vendor/pkg|src` I linked everything but tests still won't run...

I wonder if I could at least figure out if setting a different probe (e.g. TCP)


Also see https://benguild.com/2018/11/11/quickstart-golang-kubernetes-grpc-tls-lets-encrypt/
Some options would be to inject the health check in gRPC / netty  or to mangle nginx config in some way to answer the check on http/2

Now the second problem

While the probe works, it looks like the backend over HTTP/2 won't 

```
io.grpc.StatusRuntimeException: UNAVAILABLE: HTTP status code 502
invalid content-type: text/html; charset=UTF-8
headers: Metadata(:status=502,content-type=text/html; charset=UTF-8,referrer-policy=no-referrer,content-length=332,date=Mon, 17 Dec 2018 09:58:11 GMT,alt-svc=clear)
DATA-----------------------------

<html><head>
<meta http-equiv="content-type" content="text/html;charset=utf-8">
<title>502 Server Error</title>
</head>
<body text=#000000 bgcolor=#ffffff>
<h1>Error: Server Error</h1>
<h2>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.</h2>
<h2></h2>
</body></html>

DATA-----------------------------
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:233)
...
```

Regarding
> Configuring the LB backend to use ESP's HTTP/2 port did not work. I had to enable the SSL port, mount the certs (same as the LB frontend) to the container (Deployment), and expose it via the Service. So the backend is configured for the HTTP/2 protocol, but points to ESP's SSL port.

Perhaps that's because gRPC downgrades HTTP/2 to h2c?
Doesn't this defeat the purpose, though? I was assuming TLS termination in GLB. Otherwise, wouldn't a TCP load balancer for 443 with a HTTP/2 enabled backend work?

Also note that while HTTP/2 is not supported, the docs mentioned it (although until a few hours the link returned a 404). 

I'm referring to this doc https://cloud.google.com/load-balancing/docs/https/ 
Particularly
>If you want to configure an HTTP(S) load balancer using HTTP/2 with Google Kubernetes Engine Ingress, see HTTP/2 for Load Balancing with Ingress.

https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-http2 <- this was broken 

The remaining problem is that it's likely the HTTP/2 backend will likely not downgrade to h2c as gRPC does. 
https://cloud.google.com/load-balancing/docs/backend-service#troubleshoot-HTTP2

**Comment by ByteFlinger** [comment]
I have been spending the good part of 2 days trying to get this to work with Google Endpoints

I was able to setup Google Endpoints -> Ingress -> NodePort service -> Deployment (Using gRPC)

I followed the document at https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-http2 and it does give me an http/2 GLB from the Ingress. I was also able to get certmanager to provision a Let's Encrypt certificate to that GLB but in the end all I get is 502 from the LB.

This is very frustrating as it seems you are unable to use Google Endpoints with gRPC if you are on GKE (Unless you choose a different LB, like an nginx ingress)

Hope this get's solved soon. Is there another issue somewhere about this which reflects the problem better (Rather than adding examples for grpc and http2)?

**Comment by clehene** [comment]
@ByteFlinger I was able to make it work. I've updated a bunch of issues with info but missed this one.

The last thing in my case was the http2 annotation. Have you seen? https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-http2

**Comment by ByteFlinger** [comment]
@clehene I was not able to get it working sadly.

Here's what I did.

On a `1.11.2-gke.25` GKE cluster

I started by deploying the grpc sample found here https://github.com/GoogleCloudPlatform/python-docs-samples/tree/master/endpoints/bookstore-grpc

The deployment looks like this

```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: grpc-glb
  namespace: test
spec:
  selector:
    matchLabels:
      app: grpc-glb
  replicas: 1
  template:
    metadata:
      labels:
        app: grpc-glb
    spec:
      containers:
      - name: bookstore
        image: gcr.io/endpointsv2/python-grpc-bookstore-server:1
        ports:
        - containerPort: 8000
```
Then I deploy the service

```
apiVersion: v1
kind: Service
metadata:
  name: grpc-glb-service
  namespace: test
  annotations:
    cloud.google.com/app-protocols: '{"http2":"HTTP2"}'
spec:
  type: NodePort
  ports:
  - name: http2
    port: 443
    targetPort: 8000
    protocol: TCP
  selector:
    app: grpc-glb
```

and then Ingress

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: grpc-glb-ingress
  namespace: test
spec:
  backend:
    serviceName: grpc-glb-service
    servicePort: 443
```

I can confirm that I do get a Google LoadBalancer with HTTP2 support (Although the health check shows yellow after a few minutes and doesn't green)

`gcloud beta compute backend-services describe k8s-be-XXXXX--82531233212b6a88f --global`

```
backends:
...
description: '{"kubernetes.io/service-name":"test/grpc-glb-service","kubernetes.io/service-port":"443","x-features":["HTTP2"]}'
...
kind: compute#backendService
loadBalancingScheme: EXTERNAL
protocol: HTTP2
...
```

But when I try to that IP using the client all I get is `Trying to connect an http1.x server`

I should point out that if I port forward to the deployed pod and run the client towards it, it works. This is the simplest deployment I could come up with (without involving Google endpoints and such)

Did I miss something?

**Comment by agau4779** [comment]
What happens when you `curl` to port 443 with your original image @ByteFlinger ? 

GCP HTTP/2 requires HTTPS as a prerequisite - see https://cloud.google.com/load-balancing/docs/https/#backend_services and https://cloud.google.com/load-balancing/docs/backend-service#troubleshoot-HTTP2

**Comment by ByteFlinger** [comment]
@agau4779 Same problem on HTTPS

The service is not listening on https itself. I expect that to be handled by the load balancer

**Comment by ByteFlinger** [comment]
@agau4779 I have now enabled TLS in my grpc endpoint and it still does not give me a healthy check :/

Do I have to configure a separate http listener from the grpc endpoint? This is the one I tried and it seemed to work but I am confused on what is and isn't needed on the

**Comment by agau4779** [comment]
Hmm - can you ssh into into your machine and ping the port the healthcheck would be pinging (`8000`) and see what the output is? `curl --http2 https://localhost:8000`

**Comment by ByteFlinger** [comment]
Sadly after running curl from inside the container it just gives me unsupported protocol

If I port-forward to it I get `curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number`

Something tells me the issue at hand here is the esp proxy which I am trying to get it to work

https://cloud.google.com/endpoints/docs/grpc/specify-proxy-startup-options

**Comment by bgetsug** [comment]
@ByteFlinger what endpoint/path do you have the ESP health check set to? By default it uses `/healthz`. There's still no way to configure the backend health check endpoint through `Ingress` until #42 is resolved. The health check created by `Ingress` has an endpoint of `/`. Can you try setting your ESP health check endpoint to `/`? (I feel like I tried this before, but it broke ESP's proxy to the gRPC service). I believe that ESP doesn't respond with an `HTTP 200` on its `/` endpoint because it's expecting authentication headers.

**Comment by bgetsug** [comment]
Also, @agau4779, I believe there are a couple distinct architectures / use cases that should be addressed in Google's documentation:

1. `Ingress-managed GCLB --> gRPC service`
2. `Ingress-managed GCLB --> ESP (Cloud Endpoints proxy) --> gRPC service`

Perhaps the title of this issue should be changed to "Add examples using gRPC (HTTP2) with and without ESP" 

Once #42 is resolved, I would expect https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-http2 to be updated with information about how to customize the backend health check.

**Comment by ByteFlinger** [comment]
@bgetsug I am uncertain on who should be answering the health check and with what protocol. The whole thing is very confusing.

It is my understanding that, as per Google's documentation, health check on the part of the esp proxy is disabled by default so I expect any request for health checking to be sent straight to the backend. Enabling health check on esp seems to make the esp answer requests on that path itself rather than forward to the backend.

I had hoped that not enabling the health check would make would make the health check requests to go to my go backend however I am uncertain if gRPC answers 200 on the root path by default (and attempts to get the status code using curl with http2 failed).

What is even more confusing is that trying this application out

https://github.com/salrashid123/gcegrpc/tree/master/gke_ingress_lb

Works well (and that's even after I disabled  all readiness probes and just kept the  gRPC port exposed). This led me to configuring TLS and gRPC health check on my go app however still no go after that.

This leads me to think the issue is either with esp proxy or something with let's encrypt setup (though I don't think so as the certificate is being generated and provisioned by cert manager on the load balancer without any issue)

I'll give a try setting up health check on `/` in esp tomorrow morning, thanks for the suggestion

**Comment by bgetsug** [comment]
@ByteFlinger, if I understand correctly, gRPC has an RPC-based health checking protocol: https://github.com/grpc/grpc/blob/master/doc/health-checking.md, which you may be able to manually configure in your service...whether this will work with a GCLB, 🤷‍♂️.  

In my deployment I am using ESP configured with its default `/healthz` endpoint, which definitely works via plain old HTTP. Unfortunately I gave up on `Ingress` and opted to use a shell script to configure my GCLB backends and health checks with the `glcoud` CLI. (My health checks point to ESP's HTTP port on the `/healthz` endpoint.) Then I use [Terraform](https://terraform.io) to deploy my GCLB. Since I have more than one gRPC+ESP service, this seemed to be the best option as using a single ESP with multiple services is [experimental](https://github.com/cloudendpoints/esp/blob/master/doc/multiple-services.md).

**Comment by ByteFlinger** [comment]
It looks like there isn't a single working example of esp + grpc + ingress out there, assuming it works at all.

I did configure gRPC health check as you pointed out on my backend application but so far it has not worked and besides my only reason for doing so right now is to get the GCLB to proxy my requests to the underlying backend and I don't think GCLB does grpc health checking.

I may just give up on Google endpoints for now. I fail to understand how they advertise support for grpc on Google endpoints yet does not seem to work with GKE which I am guessing quite a lot of people using Google Cloud are using.

I got a couple more things to try tomorrow but don't really have much expectation they will work at this point

**Comment by bgetsug** [comment]
I can certainly echo @ByteFlinger's frustrations. One would hope that all these Google-made technologies would play nice together in a more cohesive and automated fashion. If I understand correctly, a K8S `Ingress` is supposed to automate the configuration of load balancers/proxies...whatever you want to call them. When used with GCE/GKE, I would expect an `Ingress` to be able to configure a GCLB with a more comprehensive set of options.

I believe that a resolution to #42 would be a good start and it would be great to hear from someone who works at Google with a status update or an explanation regarding the lack of progress. @agau4779, @nicksardo, @bowei...anyone? Hello, nice to meet you! 

I'm guessing that like me, other frustrated customers don't understand the ins and outs of K8S CRDs (custom resource definitions), nor do we have the time to delve so far under the hood to make this work. We just want to deploy our ESP-fronted gRPC services to GKE and have an `Ingress` automate the work necessary to configure GCLB accordingly.

If it wasn't for Cloud Endpoints and ESP, we would at the very least, have to manually "wire in" authentication middleware that supports GCP/Firebase into our gRPC services. The combined benefits of gRPC, Cloud Endpoints, and GKE are pretty tremendous. We just need a bit more simplicity from a configuration standpoint to make these combined technologies truly compelling.

I don't think it's entirely reasonable to expect someone from the community to submit a PR. I'd like to ask Google to please take this issue a bit more seriously, and perhaps get the team responsible for Ingress GCE to work with the team that's responsible for ESP (assuming they're separate) to ensure that this use case is addressed and well documented. (Also see https://github.com/GoogleCloudPlatform/endpoints-samples/issues/52# for the ESP-centric conversation.) Thank you :)

**Comment by ByteFlinger** [comment]
Well put.

Honestly right now I would be happy with a simple answer in the form of "Unfortunately it is not possible to setup Google endpoints + Ingress + Grpc right now"

**Comment by seletskiy** [comment]
@bgetsug Hey! Can you share script that you use to configure it?

**Comment by ByteFlinger** [comment]
If anybody lands here and is still looking for a way to configure GKE Ingress to create a GLB + Google endpoints (using esp) see here

https://github.com/GoogleCloudPlatform/endpoints-samples/issues/52#issuecomment-454387373

I got it to work without having to manually configure things and documented it on the link above

**Comment by rramkumar1** [comment]
Hi everyone, we now have tons of documentation on gRPC + Ingress (+ ESP optionally)

1. https://medium.com/google-cloud/gke-grpc-ingress-loadbalancing-4b9cdbc09758
2. https://cloud.google.com/solutions/exposing-grpc-services-using-cloud-endpoints-pt1
3. https://cloud.google.com/solutions/exposing-grpc-services-using-cloud-endpoints-pt2

Note that support for gRPC is not yet as first-class as it should be. We are working on that but these docs should be a good stopgap for you all.

**Comment by clehene** [comment]
@alethenorio sorry, I just saw your questions now I wish I could have helped you earlier. I'm glad you figured it out eventually.

**Comment by unludo** [comment]
Hi @rramkumar1, @salrashid123,
Actually to integrate letsencrypt with envoy has changed. Now SDS is available in Istio to correctly manage cert update, without restarting envoy.
I did not find documentation on how to configure this with GCE/GRPC.
That would be very helpful. Do you know if Google will provide more documentation on this?
thx

**Comment by salrashid123** [comment]
@unludo 

you want to configure gce+grpc on compute engine direct (vs ingress-gke context per this issue)?  if so, the following steps details how to setup grpc serivce on plain COS images as a MIG
- https://github.com/salrashid123/gcegrpc/tree/master/gce
just substitue the place towards the end where a self-signed cert is setup and use
https://cloud.google.com/load-balancing/docs/ssl-certificates  (feel free to follwup on github cited there; i'd be happy to update my sample with letsencrypt if you got it working)

**Comment by gnarea** [comment]
@rramkumar1, can we please reopen this? I don't see how the three links you pasted solved this issue. https://medium.com/google-cloud/gke-grpc-ingress-loadbalancing-4b9cdbc09758 is the closest thing to an actual example, but not quite: It still doesn't offer an actual code example of how to deal with the health checks, which is the blocker to get gRPC services to work.

Based on that article and other stuff I've read, I guess the GCE Ingress still doesn't support gRPC (because of the health checks), so we have to use one of the following workarounds if we really want to use this ingress:

- Use TCP health checks, which requires #42 and isn't a good idea anyway because they don't tell you if the app is actually healthy.
- Create a new server-side app that exposes a single HTTP endpoint, which behind the scenes pings the gRPC service.

Is that correct? If so, where can I see working examples of either?

**Comment by rramkumar1** [comment]
@gnarea Now that we are soon launching configurable health checks via BackendConfig configuring gRPC apps should be much simpler. I'll open this again so that we can get some better examples published.

**Comment by rramkumar1** [comment]
Hello everyone, it will take us a couple weeks to write and publish official docs on this subject. In the meantime, I do want to unblock folks here so look out for an unofficial guide here early next week.

**Comment by rramkumar1** [comment]
Hello folks, as promised I have some material for you all. To set expectations, the HealthCheck section of BackendConfig is only available in GKE versions 1.17.6-gke.11 and above. The current GCP documentation for this new feature lives [here](https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-features#direct_health).

Now for the fun stuff. We do not have support for gRPC health checks yet so for now we are stuck with HTTP-like health checks. What this means is that you have two options when it comes to setting up health checks for your gRPC apps:

1. Run a proxy on a separate port that translates HTTP health check requests into gRPC and then forwards the transformed request to the application port that serves a gRPC health check.
2. Run a dedicated health checker on a separate port that speaks HTTP. This health checker responds to health check requests directly rather than proxying them. 

For now, I will simply paste below the BackendConfig & Services spec that would work in both cases. I will leave it to you all to figure out how the Deployment should look like for your particular case (although I'm happy to provide more clarity if needed).

In either case, with InstanceGroup backends, this YAMLs look something like below

```
type: Service
spec:
  type: NodePort
  ports:
  - name: app
    port: 80
    targetPort: 8080
    nodePort: 30000
  - name: proxy/hc
     port: 8081
     targetPort: 8081
     nodePort: 30001
---
type: BackendConfig
spec:
  healthCheck:
    # Note that this is the NodePort for the proxy/health checker target port.
    port: 30001
    protocol: "HTTP"
    requestPath: [PATH]
```

If using a proxy with NEG backends, this YAMLs look something like below. Note in the NEG case, you do not need to expose the proxy's port in the Service. However, we have a bug related to firewall rule ports that we are working on fixing. In the meantime, refer [here](https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-features#expandable-2) for the workaround.

```
type: Service
spec:
  ports:
  - name: app
    port: 80
    targetPort: 8080
---
type: BackendConfig
spec:
  healthCheck:
    # Note that this should be the container port on the Deployment for the proxy/health checker.
    port: 8081
    protocol: "HTTP"
    requestPath: [PATH]
```

I know some folks also use ESP, so take a look at this [gist](https://gist.github.com/salrashid123/0c56c0b4f70e9f9b2fe53f844480219c) which creates a custom nginx.conf and see if that helps.

**Comment by salrashid123** [comment]
option 1 is the mux or envy cited here:
     https://github.com/salrashid123/gcegrpc/tree/master/gke_ingress_lb

option 2 with the proxy is describe here:
  https://github.com/salrashid123/grpc_health_proxy#kubernetes-pod-healthcheck

i'll update the link on option 1 to now mention the proxy.  i'd recommend option 2 since you don't have to alter your application code to account for a mux.

**Comment by gnarea** [comment]
Thanks for the updates, @rramkumar1 and @salrashid123! And apologies for the late reply.

Unfortunately, I still haven't managed to get this to work using a NEG backend for the gRPC service and @salrashid123's `grpc_health_proxy`: The backend is still marked as unhealthy.

Here's the relevant k8s resources:

```
$ kubectl get deploy gw-test-relaynet-internet-gateway-cogrpc -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "21"
    meta.helm.sh/release-name: gw-test
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2020-06-29T14:21:27Z"
  generation: 21
  labels:
    app.kubernetes.io/instance: gw-test
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
    app.kubernetes.io/version: 1.6.0
    helm.sh/chart: relaynet-internet-gateway-0.1.0
  name: gw-test-relaynet-internet-gateway-cogrpc
  namespace: default
  resourceVersion: "44837713"
  selfLink: /apis/apps/v1/namespaces/default/deployments/gw-test-relaynet-internet-gateway-cogrpc
  uid: f8c67784-e2ed-4f56-89cc-7763f1adf059
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: gw-test
      app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: gw-test
        app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
    spec:
      containers:
      - command:
        - node
        - build/main/bin/cogrpc-server.js
        env:
        - name: COGRPC_ADDRESS
          value: https://cogrpc-test.relaycorp.tech
        - name: GRPC_VERBOSITY
          value: DEBUG
        - name: GRPC_TRACE
          value: call_error,connectivity_state,http,health_check_client,http2_stream_state,http1
        envFrom:
        - configMapRef:
            name: gw-test-relaynet-internet-gateway
        - configMapRef:
            name: gw-test-relaynet-internet-gateway-cogrpc
        - secretRef:
            name: gw-test-relaynet-internet-gateway
        image: ghcr.io/relaycorp/relaynet-internet-gateway:1.6.0
        imagePullPolicy: IfNotPresent
        name: cogrpc
        ports:
        - containerPort: 8080
          name: grpc
          protocol: TCP
        resources: {}
        securityContext: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - command:
        - /bin/grpc_health_proxy
        - -http-listen-addr
        - 0.0.0.0:8081
        - -grpcaddr
        - 127.0.0.1:8080
        - -service-name
        - relaynet.cogrpc.CargoRelay
        - -v
        - "10"
        image: salrashid123/grpc_health_proxy:1.0.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /
            port: health-check
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: cogrpc-health-check
        ports:
        - containerPort: 8081
          name: health-check
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /
            port: health-check
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: gw-test-relaynet-internet-gateway-cogrpc
      serviceAccountName: gw-test-relaynet-internet-gateway-cogrpc
      shareProcessNamespace: true
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-09-02T02:08:03Z"
    lastUpdateTime: "2020-09-02T02:08:03Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2020-09-16T17:09:43Z"
    lastUpdateTime: "2020-09-16T17:30:21Z"
    message: ReplicaSet "gw-test-relaynet-internet-gateway-cogrpc-669846ff46" has
      successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 21
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

$ kubectl get backendconfig cogrpc -o yaml
apiVersion: cloud.google.com/v1
kind: BackendConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cloud.google.com/v1","kind":"BackendConfig","metadata":{"annotations":{},"labels":{"project":"public-gw"},"name":"cogrpc","namespace":"default"},"spec":{"healthCheck":{"checkIntervalSec":20,"healthyThreshold":1,"port":32578,"portSpecification":"USE_FIXED_PORT","timeoutSec":1,"type":"HTTP","unhealthyThreshold":3}}}
  creationTimestamp: "2020-07-01T10:41:21Z"
  generation: 2
  labels:
    project: public-gw
  name: cogrpc
  namespace: default
  resourceVersion: "3988438"
  selfLink: /apis/cloud.google.com/v1/namespaces/default/backendconfigs/cogrpc
  uid: 57c6219f-91ee-4e45-9576-a817c958dc3c
spec:
  healthCheck:
    checkIntervalSec: 20
    healthyThreshold: 1
    port: 32578
    timeoutSec: 1
    type: HTTP
    unhealthyThreshold: 3

$ kubectl get svc gw-test-relaynet-internet-gateway-cogrpc -o yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    beta.cloud.google.com/backend-config: '{"ports":{"grpc":"cogrpc"}, "default":
      "cogrpc"}'
    cloud.google.com/app-protocols: '{"grpc":"HTTP2"}'
    cloud.google.com/neg: '{"ingress": true}'
    cloud.google.com/neg-status: '{"network_endpoint_groups":{"8081":"k8s1-4ddb902c-defaul-gw-test-relaynet-internet-gate-80-01fab572"},"zones":["europe-west2-a"]}'
    meta.helm.sh/release-name: gw-test
    meta.helm.sh/release-namespace: default
    service.alpha.kubernetes.io/app-protocols: '{"grpc":"HTTP2"}'
  creationTimestamp: "2020-06-29T14:21:27Z"
  labels:
    app.kubernetes.io/instance: gw-test
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
    app.kubernetes.io/version: 1.6.0
    helm.sh/chart: relaynet-internet-gateway-0.1.0
  name: gw-test-relaynet-internet-gateway-cogrpc
  namespace: default
  resourceVersion: "44927024"
  selfLink: /api/v1/namespaces/default/services/gw-test-relaynet-internet-gateway-cogrpc
  uid: 9e28d795-f1ee-49ab-a307-1424b016d46a
spec:
  clusterIP: 10.16.5.230
  externalTrafficPolicy: Cluster
  ports:
  - name: grpc
    nodePort: 32112
    port: 8081
    protocol: TCP
    targetPort: grpc
  - name: health-check
    nodePort: 32758
    port: 8082
    protocol: TCP
    targetPort: health-check
  selector:
    app.kubernetes.io/instance: gw-test
    app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
  sessionAffinity: None
  type: NodePort
status:
  loadBalancer: {}

$ kubectl get ing gw-test-relaynet-internet-gateway -o yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.gcp.kubernetes.io/pre-shared-cert: mcrt-c4ce565e-27cb-4d5b-a41c-56e9c926d2bf
    ingress.kubernetes.io/backends: '{"k8s-be-30803--4ddb902ca0d0fd46":"HEALTHY","k8s1-4ddb902c-defaul-gw-test-relaynet-internet-gate-80-01fab572":"Unknown","k8s1-4ddb902c-defaul-gw-test-relaynet-internet-gate-80-614f02b3":"HEALTHY"}'
    ingress.kubernetes.io/https-forwarding-rule: k8s2-fs-vfqim3gh-default-gw-test-relaynet-internet-gat-s1dm076y
    ingress.kubernetes.io/https-target-proxy: k8s2-ts-vfqim3gh-default-gw-test-relaynet-internet-gat-s1dm076y
    ingress.kubernetes.io/ssl-cert: mcrt-c4ce565e-27cb-4d5b-a41c-56e9c926d2bf
    ingress.kubernetes.io/url-map: k8s2-um-vfqim3gh-default-gw-test-relaynet-internet-gat-s1dm076y
    kubernetes.io/ingress.allow-http: "false"
    kubernetes.io/ingress.global-static-ip-name: gateway-example-managed-cert
    meta.helm.sh/release-name: gw-test
    meta.helm.sh/release-namespace: default
    networking.gke.io/managed-certificates: gateway-example
  creationTimestamp: "2020-06-29T14:21:27Z"
  finalizers:
  - networking.gke.io/ingress-finalizer-V2
  generation: 2
  labels:
    app.kubernetes.io/instance: gw-test
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: relaynet-internet-gateway
    app.kubernetes.io/version: 1.6.0
    helm.sh/chart: relaynet-internet-gateway-0.1.0
  name: gw-test-relaynet-internet-gateway
  namespace: default
  resourceVersion: "44927162"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/gw-test-relaynet-internet-gateway
  uid: 58b09753-338c-406b-908b-068dab28e47c
spec:
  rules:
  - host: pohttp-test.relaycorp.tech
    http:
      paths:
      - backend:
          serviceName: gw-test-relaynet-internet-gateway-pohttp
          servicePort: http
  - host: cogrpc-test.relaycorp.tech
    http:
      paths:
      - backend:
          serviceName: gw-test-relaynet-internet-gateway-cogrpc
          servicePort: grpc
  tls:
  - hosts:
    - pohttp-test.relaycorp.tech
    - cogrpc-test.relaycorp.tech
status:
  loadBalancer:
    ingress:
    - ip: 35.201.77.144
```

And here's the firewall rule I created as a workaround for the bug with NEG backends:

```
$ gcloud compute --project=public-gw firewall-rules describe neg-backend-workaround
allowed:
- IPProtocol: tcp
  ports:
  - '8081'
creationTimestamp: '2020-09-16T11:06:32.242-07:00'
description: ''
direction: INGRESS
disabled: false
id: '5022437837888285383'
kind: compute#firewall
logConfig:
  enable: false
name: neg-backend-workaround
network: https://www.googleapis.com/compute/v1/projects/public-gw/global/networks/default
priority: 1000
selfLink: https://www.googleapis.com/compute/v1/projects/public-gw/global/firewalls/neg-backend-workaround
sourceRanges:
- 130.211.0.0/22
- 35.191.0.0/16
```

(I copy/pasted the `sourceRanges` from the article that @rramkumar1 shared)

I also confirmed that the gRPC proxy is working as intended: I got the following `200 OK` whilst doing port forwarding (`kubectl port-forward pods/gw-test-relaynet-internet-gateway-cogrpc-669846ff46-cnz6z 8081:8081`)

```
$ curl -v http://localhost:8081
*   Trying 127.0.0.1:8081...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8081 (#0)
> GET / HTTP/1.1
> Host: localhost:8081
> User-Agent: curl/7.65.3
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 16 Sep 2020 21:23:53 GMT
< Content-Length: 34
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host localhost left intact
relaynet.cogrpc.CargoRelay SERVING
```

Finally, here's a sample log from the NEG:

```
{
   "insertId": "1ndoqi7c10m",
   "jsonPayload": {
     "healthCheckProbeResult": {
       "detailedHealthState": "TIMEOUT",
       "healthState": "UNHEALTHY",
       "probeSourceIp": "35.191.9.34",
       "targetPort": 32578,
       "previousDetailedHealthState": "UNKNOWN",
       "healthCheckProtocol": "HTTP",
       "ipAddress": "10.12.4.13",
       "probeCompletionTimestamp": "2020-09-16T21:15:45.637224925Z",
       "probeRequest": "/",
       "probeResultText": "HTTP response: , Error: Connection refused",
       "responseLatency": "0.001627s",
       "targetIp": "10.12.4.13",
       "previousHealthState": "UNHEALTHY"
     }
   },
   "resource": {
     "type": "gce_network_endpoint_group",
     "labels": {
       "zone": "europe-west2-a",
       "network_endpoint_group_id": "6684639526294893251",
       "project_id": "public-gw"
     }
   },
   "timestamp": "2020-09-16T21:15:46.150817444Z",
   "severity": "INFO",
   "logName": "projects/public-gw/logs/compute.googleapis.com%2Fhealthchecks",
   "receiveTimestamp": "2020-09-16T21:15:46.964901383Z"
 }
```

... Which could probably be caused by the bug @rramkumar1 mentioned, involving NEG backends.

Any idea what could be going wrong?

**Comment by gnarea** [comment]
Sorry, just realised the `BackendConfig` wasn't being applied because `protocol: HTTP` should be `type: HTTP`. Having fixed that, the backend is now marked as healthy! :tada:

**Comment by gnarea** [comment]
Unfortunately, the health check is broken again after I made a [trivial change to the deployment](https://github.com/relaycorp/relaynet-internet-gateway-chart/commit/8d4e08a9cf9751eb0785965ce8f874c30725a791). Here's what I observed:

1. The new version of the deployment got stuck for a while because the BackendConfig heath check didn't pass for the new pod.
1. After a while, the new deployment took over but the health check remains broken.

Here's the `BackendConfig` right now:

```
$ kubectl get backendconfig cogrpc -o yaml
apiVersion: cloud.google.com/v1
kind: BackendConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"cloud.google.com/v1","kind":"BackendConfig","metadata":{"annotations":{},"labels":{"project":"public-gw"},"name":"cogrpc","namespace":"default"},"spec":{"healthCheck":{"port":8082,"requestPath":"/","type":"HTTP"}}}
  creationTimestamp: "2020-07-01T10:41:21Z"
  generation: 4
  labels:
    project: public-gw
  name: cogrpc
  namespace: default
  resourceVersion: "44943939"
  selfLink: /apis/cloud.google.com/v1/namespaces/default/backendconfigs/cogrpc
  uid: 57c6219f-91ee-4e45-9576-a817c958dc3c
spec:
  healthCheck:
    port: 8082
    requestPath: /
    type: HTTP
```

I haven't changed anything else. Any idea what could be going on?

**Comment by gnarea** [comment]
It seems like the issue above was transient. After an hour or so, and without making any further changes, the backend became healthy again. I've also been doing further deployments and they've been applied successfully very quickly.

I'm facing a new issue though: The LB seems to be using the wrong protocol when talking to my backend. Here's what I get in the gRPC server logs when I enable tracing:

```
ipv4:35.191.1.73:52717: Complete BDP ping err={"created":"@1600333009.483227156","description":"Failed parsing HTTP/2","file":"../deps/grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc","file_line":2582,"referenced_errors":[{"created":"@1600333009.483221089","description":"Connect string mismatch: expected 'P' (80) got '\u0016' (22) at byte 0","file":"../deps/grpc/src/core/ext/transport/chttp2/transport/parsing.cc","file_line":97}]}
```

I'm trying to do TLS termination at the LB, so that the LB and the gRPC service can communicate over HTTP/2 without TLS (aka HTTP/2 over TCP or "h2c"). And **I can make requests to my gRPC service if I skip the LB and use `kubectl port-forward`**.

I wonder whether the following annotations on the service are the source of the problem:

```
    cloud.google.com/app-protocols: '{"grpc":"HTTP2"}'
    service.alpha.kubernetes.io/app-protocols: '{"grpc":"HTTP2"}'
```

... Does "HTTP2" imply "HTTP/2 with TLS"? If so, how can I keep HTTP2 but disable TLS?

**Comment by rramkumar1** [comment]
@gnarea Thanks for working through this!

One important point to note is that communication between the LB and your service must use TLS if using HTTP2. It is a hard requirement. You may use a self-signed cert on your service if you choose and it will work just fine.

**Comment by gnarea** [comment]
Thanks @rramkumar1! That must be it. I'll try that

**Comment by gnarea** [comment]
@rramkumar1, since presumably anyone deploying gRPC with this ingress will be on the same boat, is there any approach in particular that you'd recommend to get TLS to work with no changes to the application?

For example, I was looking for a ready-made Docker image that would self-issue an ephemeral TLS certificate at startup time and it'd proxy my HTTP2/gRPC service, but I couldn't find anything. I found [this](https://github.com/mnuessler/docker-tls-termination-proxy) but it's unmaintained.

Of course, I could make my application generate the TLS cert and use it in the gRPC server, but there's got to be an easier/standard way to do this without changing the app, right?

**Comment by rramkumar1** [comment]
@gnarea That's an interesting question but I don't have a great answer for you there. I don't see a way to do this without tweaking your app.

**Comment by bowei** [comment]
It seems like unencrypted HTTP2 is not as common -- https://http2.github.io/faq/#does-http2-require-encryption, so that is probably why support for it on the backend is not possible right now.

Generating the self-signed cert is straightforward, can be done with an init container with a script, but you really have to watch out for cert expiry.

```
#!/bin/bash
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365
```

**Comment by gnarea** [comment]
Thanks @rramkumar1 and @bowei!

Putting [a TLS reverse proxy](https://github.com/zazukoians/docker-hitch) in front of the gRPC server fixed that problem, but it's uncovered another problem: The load balancer is failing to connect to the proxy and I suspect it's because the LB is refusing the self-issued certificate (not necessarily because it's self-issued, but for a different reason). Here's what's happening:

1. I make a gRPC request to the LB over the Internet. Behind the scenes, this is an HTTP/2 POST request to `https://cogrpc-test.relaycorp.tech/relaynet.cogrpc.CargoRelay/DeliverCargo`. The result is a 502 response.
1. In the LB logs, I can see that the reason for the failure is `statusDetails: "failed_to_connect_to_backend"`:
   ```
    {
    "jsonPayload": {
        "statusDetails": "failed_to_connect_to_backend",
        "@type": "type.googleapis.com/google.cloud.loadbalancing.type.LoadBalancerLogEntry"
    },
    "httpRequest": {
        "requestMethod": "POST",
        "requestUrl": "https://cogrpc-test.relaycorp.tech/relaynet.cogrpc.CargoRelay/DeliverCargo",
        "requestSize": "526587",
        "status": 502,
        "responseSize": "447",
        "userAgent": "grpc-node/1.24.2 grpc-c/8.0.0 (linux; chttp2; ganges)",
        "remoteIp": "<redacted>",
        "serverIp": "10.12.2.9",
        "latency": "9.005462s"
    },
    "resource": {
        "type": "http_load_balancer",
        ...
        }
    },
    "timestamp": "2020-09-17T20:49:31.046721Z",
    "severity": "WARNING",
    "logName": "projects/public-gw/logs/requests",
    "receiveTimestamp": "2020-09-17T20:49:40.415675057Z",
    }
   ```
1. In the TLS proxy logs, the request above maps to the following logs -- which suggests the TLS handshake failed:
   ```
    E 2020-09-17T20:49:31.344735524Z 20200917T204931.344037 [   40] 10.12.2.1:52170 :0 7:8 ssl handshake start
    E 2020-09-17T20:49:31.344918961Z 20200917T204931.344054 [   40] 10.12.2.1:52170 :0 7:8 ssl client handshake revents=1
    E 2020-09-17T20:49:31.345076960Z 20200917T204931.344089 [   40] 10.12.2.1:52170 :0 7:8 ssl client handshake err=SSL_ERROR_SYSCALL
    E 2020-09-17T20:49:31.345232699Z 20200917T204931.344097 [   40] {client} SSL socket error in handshake: No error information
   ```
1. As expected, there are no corresponding logs on the gRPC server (with tracing enabled).
1. I can confirm the TLS proxy is configured with a valid (yet untrusted) certificate and that HTTP/2 ALPN is supported. Here's what I get when I `curl` the proxy using port-forwarding in k8s (note that this isn't a valid gRPC request):
   ```
    $ curl -k -v https://localhost:8080
    *   Trying 127.0.0.1:8080...
    * TCP_NODELAY set
    * Connected to localhost (127.0.0.1) port 8080 (#0)
    * ALPN, offering h2
    * ALPN, offering http/1.1
    * successfully set certificate verify locations:
    *   CAfile: none
    CApath: /etc/ssl/certs
    * TLSv1.3 (OUT), TLS handshake, Client hello (1):
    * TLSv1.3 (IN), TLS handshake, Server hello (2):
    * TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
    * TLSv1.3 (IN), TLS handshake, Certificate (11):
    * TLSv1.3 (IN), TLS handshake, CERT verify (15):
    * TLSv1.3 (IN), TLS handshake, Finished (20):
    * TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
    * TLSv1.3 (OUT), TLS handshake, Finished (20):
    * SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
    * ALPN, server accepted to use h2
    * Server certificate:
    *  subject: C=CH; ST=Zurich; L=Zurich; O=Snakeoil Inc; OU=IT Department; CN=10.12.2.9
    *  start date: Sep 17 21:38:37 2020 GMT
    *  expire date: Sep 17 21:38:37 2021 GMT
    *  issuer: C=CH; ST=Zurich; L=Zurich; O=Snakeoil Inc; OU=IT Department; CN=10.12.2.9
    *  SSL certificate verify result: self signed certificate (18), continuing anyway.
    * Using HTTP2, server supports multi-use
    * Connection state changed (HTTP/2 confirmed)
    * Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
    * Using Stream ID: 1 (easy handle 0x55f78a0561d0)
    > GET / HTTP/2
    > Host: localhost:8080
    > User-Agent: curl/7.65.3
    > Accept: */*
    > 
    * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
    * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
    * old SSL session ID is stale, removing
    * Connection state changed (MAX_CONCURRENT_STREAMS == 3)!
    * HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
    * stopped the pause stream!
    * Connection #0 to host localhost left intact
    curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
   ```

How can I tell why the LB is aborting during the TLS handshake? Are there any logs I can check?

Is the LB expecting the backend certificate to have a specific Common Name or Subject Alternative Name? I'm currently using the IP of the pod (`10.12.2.9` as shown in the logs above).

**Comment by rramkumar1** [comment]
@gnarea Is your backend trying to do mTLS? If so, its not supported on GCLB.

Your errors look very similar to other errors I've seen where folks are trying to use mTLS.

**Comment by gnarea** [comment]
@rramkumar1, no, no mutual TLS.

I've managed to solve this issue by ditching the TLS proxy and getting the gRPC server to do TLS with a self-issued certificate.

I think the problem is that the TLS proxy was generating a self-issued certificate with a CommonName (set to the IP address of the pod), but it wasn't setting the SubjectAlternativeName extension. **This is important**. So now I have some custom code to create the certificate and set the SAN:

```
# TypeScript code using the "selfsigned" NPM package
# I need to check if the "keyUsage" extension is required by the LB or the Go gRPC client used by the health checks
# ipAddress should be the IP of the pod

async function selfIssueCertificate(ipAddress: string): Promise<KeyCertPair> {
  const pems = selfsigned.generate([{ name: 'commonName', value: ipAddress }], {
    days: 365,
    extensions: [
      { name: 'basicConstraints', cA: true },
      {
        dataEncipherment: true,
        digitalSignature: true,
        keyCertSign: true,
        keyEncipherment: true,
        name: 'keyUsage',
        nonRepudiation: true,
      },
      {
        altNames: [
          {
            ip: ipAddress,
            type: 7, // IP Address
          },
        ],
        name: 'subjectAltName',
      },
    ],
  });
  return { cert_chain: Buffer.from(pems.cert), private_key: Buffer.from(pems.private) };
}
```

I can't believe this is finally working.

GCP folks, some feedback:

- **Please** log the reason why the connection failed. `failed_to_connect_to_backend` is not enough at all.
- Please create a Docker image that acts as a TLS proxy with all the appropriate configuration to work with your LB and an HTTP/2 backend.

**Comment by gnarea** [comment]
Here's the final deployment in case it helps anyone:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gw-test-relaynet-internet-gateway-cogrpc
  labels:
    helm.sh/chart: relaynet-internet-gateway-0.1.0
    app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
    app.kubernetes.io/instance: gw-test
    app.kubernetes.io/version: "1.6.3"
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
      app.kubernetes.io/instance: gw-test
  template:
    metadata:
      labels:
        app.kubernetes.io/name: relaynet-internet-gateway-cogrpc
        app.kubernetes.io/instance: gw-test
    spec:
      containers:
        - name: cogrpc
          image: "<YOU-IMAGE>"
          imagePullPolicy: IfNotPresent
          env:
            - name: SERVER_IP_ADDRESS
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          ports:
            - name: grpc
              containerPort: 8080
              protocol: TCP
        - name: cogrpc-health-check
          image: salrashid123/grpc_health_proxy:1.0.0
          imagePullPolicy: IfNotPresent
          command:
            - "/bin/grpc_health_proxy"
            - "-http-listen-addr"
            - "0.0.0.0:8082"
            - "-grpcaddr"
            - "127.0.0.1:8080"
            - "-service-name"
            - "relaynet.cogrpc.CargoRelay"
            - "-grpctls"
            - "-grpc-tls-no-verify"
            - "--logtostderr=1"
            - "-v"
            - "10"
          ports:
            - name: health-check
              containerPort: 8082
              protocol: TCP
          livenessProbe:
            httpGet:
              port: "health-check"
          readinessProbe:
            httpGet:
              port: "health-check"

```

**Comment by raboof** [comment]
> Please create a Docker image that acts as a TLS proxy with all the appropriate configuration to work with your LB and an HTTP/2 backend

While this would be a useful component in the short term, I'd say the 'real' fix would be to allow using h2c (HTTP/2 without HTTPS) between the loadbalancer and the service 'directly'.

**Comment by gnarea** [comment]
I totally agree with @raboof: GCP LBs should be able to handle TLS termination regardless of whether the backend uses HTTP 1 or 2.

Here's an issue to improve the logging in case anyone's interested: https://issuetracker.google.com/issues/168884858

**Comment by bowei** [comment]
We will send this to the L7 proxy folks -- in the meantime, comment on the issue tracker -- it's a good way to show there is interest as well.

**Comment by salrashid123** [comment]
@gnarea you can use envoy to proxy the tls intended for the gRPC service.  [here](https://github.com/salrashid123/gcegrpc/tree/master/app#using-envoy-for-tls-proxy) is an exampleof that i just updated plus an easy button way to gen sni from subordinate cert.  
however, If you  use envoy now, its yet  another container (making a grand total of containers `+2 (envoy+grpc_healthcheck binary)` just for grpc lb which is crazy complex...).    You can, if you really, rally want to, mash lua to help with the healthchecks and still proxy (meaning you now down to `+1`  but now have complexity shunted to the envoy config.. [here](https://github.com/salrashid123/gcegrpc/tree/master/gke_ingress_lb/gke_ingress_lb_backend_config/envoy_proxy).  these are workarounds.  there needs to be an easier way

**Comment by gnarea** [comment]
> these are workarounds. there needs to be an easier way

Indeed.

BTW, I've just created a separate issue for the gRPC health checks: https://issuetracker.google.com/issues/168994852

**Comment by gnarea** [comment]
@bowei, is there a issue for the TLS termination on the L7 proxy? I'd like to star it and get updates.

**Comment by bowei** [comment]
@gnarea I just noticed that this issue is pretty old and not specific to support for unencrypted backends for HTTP2. It probably is better to open an issue for that topic so the activity can be tracked easier.

**Comment by gnarea** [comment]
Done @bowei: https://issuetracker.google.com/issues/169122105

**Comment by override80** [comment]
For all of you coming here as of today, this is what I did to get a grpc plaintext application working on GKE behind GCLB. Thanks @salrashid123 and @gnarea for providing insights. I used the docker-hitch sidecar from [here](https://github.com/zazukoians/docker-hitch/) instead of envoy (much simpler to me) which generates TLS certificates on the fly picking up pod ip address at startup. I slightly modified it to accomodate http2 and alpn with the correct ciphers supported by GCLB ([here](https://github.com/override80/docker-hitch) the fork, it might be better to move params to Dockerfile variable, but anyway...).

Here you can see redacted manifests for k8s:

`apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/component: component
    app.kubernetes.io/instance: istance_name
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app_name
    app.kubernetes.io/part-of: part_name
    app.kubernetes.io/version: 083d4037
  name: app_name
  namespace: ns_name
spec:
  minReadySeconds: 5
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 5
  selector:
    matchLabels:
      app.kubernetes.io/instance: istance_name
      app.kubernetes.io/name: app_name
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      labels:
        app.kubernetes.io/component: component
        app.kubernetes.io/instance: istance_name
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: app_name
        app.kubernetes.io/part-of: part_name
        app.kubernetes.io/version: 083d4037
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchExpressions:
                - key: app.kubernetes.io/name
                  operator: In
                  values:
                  - app_name
              topologyKey: kubernetes.io/hostname
            weight: 100
      containers:
      - image: override80/hitch:1.0.0
        imagePullPolicy: IfNotPresent
        env:
        - name: MY_POD_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.podIP
        name: hitch
        ports:
        - containerPort: 8443
          name: https
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - image: gcr.io/cloud-solutions-images/grpc_health_proxy
        imagePullPolicy: IfNotPresent
        args:
        - --http-listen-addr=0.0.0.0:8080
        - --grpcaddr=localhost:8000
        - --service-name=app.Check
        - --logtostderr=1
        - -v=1
        name: hc-proxy
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - image: redacted.registry/app/server:e0dfbce1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
        name: app
        ports:
        - containerPort: 8000
          name: grpc
          protocol: TCP
        - containerPort: 9091
          name: prom
          protocol: TCP
        resources:
          limits:
            cpu: 512m
            memory: 128Mi
          requests:
            cpu: 125m
            memory: 64Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /app/.env
          name: configs
          readOnly: true
          subPath: .env
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: registry-credentials
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: configs
        secret:
          defaultMode: 420
          secretName: configs
`

`
apiVersion: cloud.google.com/v1
kind: BackendConfig
metadata:
  name: app-grpc-backendconfig
  namespace: ns_name
spec:
  healthCheck:
    port: 8080
    requestPath: /healthz
    type: HTTP
`

`
apiVersion: v1
kind: Service
metadata:
  annotations:
    cloud.google.com/app-protocols: '{"https":"HTTP2"}'
    cloud.google.com/backend-config: '{"default": "app-grpc-backendconfig"}'
    cloud.google.com/neg: '{"ingress": true}'
    prometheus.io/scrape: "true"
  labels:
    app.kubernetes.io/component: component
    app.kubernetes.io/instance: istance_name
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app_name
    app.kubernetes.io/part-of: part_name
    app.kubernetes.io/version: 083d4037
  name: svc_name
  namespace: ns_name
spec:
  externalTrafficPolicy: Local
  ports:
  - name: https
    port: 8443
    protocol: TCP
    targetPort: 8443
  - name: http
    port: 8080
    protocol: TCP
    targetPort: 8080
  - name: grpc
    port: 80
    protocol: TCP
    targetPort: grpc
  - name: prom
    port: 9091
    protocol: TCP
    targetPort: prom
  selector:
    app.kubernetes.io/instance: istance_name
    app.kubernetes.io/name: app_name
  sessionAffinity: None
  type: NodePort
`

`
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    acme.cert-manager.io/http01-edit-in-place: "true"
    cert-manager.io/cluster-issuer: production
    external-dns.alpha.kubernetes.io/hostname: extarnal.dns.name
    external-dns.alpha.kubernetes.io/ttl: "200"
    kubernetes.io/ingress.class: gce
  labels:
    app.kubernetes.io/component: component
    app.kubernetes.io/instance: istance_name
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app_name
    app.kubernetes.io/part-of: part_name
    app.kubernetes.io/version: 083d4037
  name: app_name
  namespace: ns_name
spec:
  backend:
    serviceName: ingress
    servicePort: https
  tls:
  - hosts:
    - extarnal.dns.name
    secretName: external-dns-name-tls
`

Hope this helps guys! Thanks to everyone for your posts. [This](https://www.loginradius.com/blog/async/service-mesh-with-envoy/) is a good resource as well.
